### ISSUE TITLE
NewRelic extension is sometimes downloaded after composer install

### ISSUE THREAD

**Issue by mjanser** [cloudfoundry/php-buildpack#210]
#### What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?
```json
{
   "name": "",
   "build": "",
   "support": "https://support.pivotal.io",
   "version": 0,
   "description": "",
   "authorization_endpoint": "https://...",
   "token_endpoint": "https://...",
   "min_cli_version": "6.7.0",
   "min_recommended_cli_version": "6.11.2",
   "api_version": "2.65.0",
   "app_ssh_endpoint": "...:2222",
   "app_ssh_host_key_fingerprint": "...",
   "app_ssh_oauth_client": "ssh-proxy",
   "routing_endpoint": "https://...",
   "logging_endpoint": "wss://...",
   "doppler_logging_endpoint": "wss://..."
}
```
`cf version 6.26.0+9c9a261.2017-04-06`

#### What version of the buildpack you are using?
Buildpack version 4.3.31-1493150241

#### If you were attempting to accomplish a task, what was it you were attempting to do?
I'm deploying a PHP application with newrelic support using `cf push`.

#### What did you expect to happen?
I expect that the NewRelic extension is always there and ready before `composer install` is called.

#### What was the actual behavior?
Sometimes the NewRelic extension is not yet downloaded when `composer install` is called. It seems that the order of some things during stage is random (or even parallel). If it happens that composer is called before the NewRelic extension was downloaded, the extension is not available at this point, but it should.

This is problem when using [EkinoNewRelicBundle](https://github.com/ekino/EkinoNewRelicBundle) for example. The bundle requires the NewRelic extension to be loaded when generating the cache.

```
...
Staging...
-------> Buildpack version 4.3.31-1493150241
Installing Nginx
Downloaded [file:///tmp/buildpacks/652caf45b6f86e88785668209fc6b1fb/dependencies/https___buildpacks.cloudfoundry.org_dependencies_nginx_nginx-1.11.13-linux-x64-0ae5f6fb.tgz] to [/tmp]
Installing PHP
PHP 7.0.18
Downloaded [file:///tmp/buildpacks/652caf45b6f86e88785668209fc6b1fb/dependencies/https___cf-buildpacks-deps.s3.amazonaws.com_dependencies_php7_php-7.0.18-linux-x64.tgz] to [/tmp]
The extension 'newrelic' is not provided by this buildpack.
Downloaded [file:///tmp/buildpacks/652caf45b6f86e88785668209fc6b1fb/dependencies/https___cf-buildpacks-deps.s3.amazonaws.com_dependencies_php7_php-7.0.18-linux-x64.tgz] to [/tmp]
Downloaded [file:///tmp/buildpacks/652caf45b6f86e88785668209fc6b1fb/dependencies/https___buildpacks.cloudfoundry.org_dependencies_composer_composer-1.4.1-0eedfb38.phar] to [/tmp]
Loading composer repositories with package information
Installing dependencies from lock file
Package operations: 0 installs, 0 updates, 0 removals
Generating autoload files
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache
 // Clearing the cache for the prod environment with debug                      
 // false                                                                       
 [OK] Cache for the "prod" environment (debug=false) was successfully cleared.  
Downloaded [file:///tmp/buildpacks/652caf45b6f86e88785668209fc6b1fb/dependencies/https___download.newrelic.com_php_agent_archive_6.3.0.161_newrelic-php5-6.3.0.161-linux.tar.gz] to [/tmp]
Finished: [2017-05-16 11:07:03.938325]
Exit status 0
...
```

**Comment by dmikusa-pivotal** [comment]
Extension order is [somewhat fixed](https://github.com/cloudfoundry/php-buildpack/blob/master/scripts/compile.py#L35-L48).  It basically goes like this:

1. Web Server
2. PHP
3. Then all the other extensions that ship w/the build pack
4. Then any extensions shipped with your app

For #3 & #4 the order within that is arbitrary.  This is because the code that bulk loads those extensions is using Python's `os.listdir` which itself defines no order.

A handful of solutions come to mind:

- sort the extensions alphabetically (this would always make Composer first)
- change [this](https://github.com/cloudfoundry/php-buildpack/blob/master/scripts/compile.py#L42-L43) to individually load build pack's.  Ex: `.extensions()..from_build_pack('extensions/sessions')`
- extend extension mechanism to allow each extension to define its own ordering then order on this

@sclevine - Your thoughts?

**Comment by mjanser** [comment]
A workaround for my use case is clearing the cache again in `ADDITIONAL_PREPROCESS_CMDS` like this in `options.json`:

```json
{
  "ADDITIONAL_PREPROCESS_CMDS": [
    "/home/vcap/app/bin/console cache:clear"
  ]
}
```

**Comment by sclevine** [comment]
I'll prioritize this as a bug, and let the engineers picking up the story decide on an implementation.

**Comment by sclevine** [comment]
Fixed on develop: d0657abfee026f877bf2402dc98e698e7f7a1a60

@mjanser can you confirm that develop addresses your issue?
(You can push with `cf push youapp -b http://github.com/cloudfoundry/php-buildpack#develop` to test.)

**Comment by mjanser** [comment]
@sclevine yes, this fixes my issue. Thank you!
